### ISSUE TITLE
Compiling with CMake in Linux dynamically links to system Lua, should do static compilation

### ISSUE THREAD

**Issue by rfkrocktk** [stepmania/stepmania#1006]
I think the problem is in here: https://github.com/stepmania/stepmania/blob/master/extern/CMakeProject-lua.cmake

I'm building on Fedora 23 and it is still linking to the system Lua in the RPM build, which I can't control. My specfile is here:

```
%global _hardened_build 1

%define package_name stepmania
%define package_version 5.0.10
%define package_release 2

Name: %{package_name}
Version: %{package_version}
Release: %{package_release}%{?dist}
Summary: StepMania is a free dance and rhythm game for Windows, Mac, and Linux.
License: MIT
Packager: Naftuli Tzvi Kay <rfkrocktk@gmail.com>
URL: http://stepmania.com
Group: Applications/Multimedia
Source: https://github.com/stepmania/stepmania/archive/v%{package_version}.tar.gz#/stepmania-%{package_version}.tar.gz
Patch0: stepmania-patch-0001-ffmpeg-pic.patch
Patch1: stepmania-patch-0002-cmake-i386.patch
Patch2: stepmania-patch-0003-use-system-libraries.patch

ExclusiveArch: %{ix86} x86_64

BuildRequires: alsa-lib-devel
BuildRequires: atk-devel
BuildRequires: autoconf
BuildRequires: automake
BuildRequires: bash
BuildRequires: binutils
BuildRequires: bzip2-devel
BuildRequires: bzip2-libs
BuildRequires: cairo-devel
BuildRequires: cmake
BuildRequires: coreutils
BuildRequires: cpio
BuildRequires: diffutils
BuildRequires: dwz
BuildRequires: gcc
BuildRequires: gcc-c++
BuildRequires: gdk-pixbuf2-devel
BuildRequires: glew-devel
BuildRequires: glib2-devel
BuildRequires: glibc
BuildRequires: glibc-common
BuildRequires: glibc-devel
BuildRequires: glibc-headers
BuildRequires: gtk2-devel
BuildRequires: jsoncpp-devel
BuildRequires: kernel-headers
BuildRequires: libX11-devel
BuildRequires: libXrandr-devel
BuildRequires: libXrender-devel
BuildRequires: libgcc
BuildRequires: libjpeg-turbo-devel
BuildRequires: libmad-devel
BuildRequires: libogg-devel
BuildRequires: libpng-devel
BuildRequires: libstdc++-devel
BuildRequires: libva-devel
BuildRequires: libvdpau-devel
BuildRequires: libvorbis-devel
BuildRequires: m4
BuildRequires: make
BuildRequires: mesa-libEGL-devel
BuildRequires: mesa-libGL-devel
BuildRequires: mesa-libGLU-devel
BuildRequires: pango-devel
BuildRequires: pcre-devel
BuildRequires: pkgconfig
BuildRequires: pulseaudio-libs-devel
BuildRequires: xorg-x11-proto-devel
BuildRequires: yasm
BuildRequires: git
BuildRequires: checksec
BuildRequires: libtomcrypt-devel
BuildRequires: libtommath-devel

Requires: gtk-update-icon-cache
Requires: hicolor-icon-theme
Requires: desktop-file-utils

%description
StepMania is a free dance and rhythm game for Windows, Mac, and Linux. It features 3D graphics, keyboard and "dance pad"
support, and an editor for creating your own steps.

%prep

%setup -q
%patch0 -p1
%patch1 -p1
%patch2 -p1
sed -i 's:Exec=stepmania:Exec=/usr/share/stepmania-5.0/stepmania:g' stepmania.desktop

%build
%cmake -G "Unix Makefiles" \
    -DCMAKE_INSTALL_PREFIX=/usr/share \
    -DCMAKE_SKIP_RPATH:BOOL=YES \
    -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=YES \
    -DENABLE_PIC:BOOL=ON \
    -DWITH_FULL_RELEASE:BOOL=ON \
    -DCMAKE_BUILD_TYPE=Release \
    .
make %{_smp_mflags}

%install
%make_install
test -d %{buildroot}%{_datadir} || mkdir %{buildroot}%{_datadir}
cp -r icons/ %{buildroot}%{_datadir}/icons
test -d %{buildroot}%{_datadir}/applications || mkdir %{buildroot}%{_datadir}/applications
install -m 0644 stepmania.desktop %{buildroot}%{_datadir}/applications/

%check
checksec --file %{buildroot}%{_datadir}/stepmania-5.0/stepmania
checksec --file %{buildroot}%{_datadir}/stepmania-5.0/GtkModule.so

%post
touch --no-create %{_datadir}/icons/hicolor &>/dev/null || :
/usr/bin/update-desktop-database &> /dev/null || :
%postun
if [ $1 -eq 0 ] ; then
    touch --no-create %{_datadir}/icons/hicolor &>/dev/null
    gtk-update-icon-cache %{_datadir}/icons/hicolor &>/dev/null || :
fi
/usr/bin/update-desktop-database &> /dev/null || :
%posttrans
gtk-update-icon-cache %{_datadir}/icons/hicolor &>/dev/null || :

%files
# binary
%{_datadir}/stepmania-5.0/stepmania
%{_datadir}/stepmania-5.0/GtkModule.so
# assets
%{_datadir}/stepmania-5.0/Announcers
%{_datadir}/stepmania-5.0/BGAnimations
%{_datadir}/stepmania-5.0/BackgroundEffects
%{_datadir}/stepmania-5.0/BackgroundTransitions
%{_datadir}/stepmania-5.0/Characters
%{_datadir}/stepmania-5.0/Courses
%{_datadir}/stepmania-5.0/Data
%{_datadir}/stepmania-5.0/Docs
%{_datadir}/stepmania-5.0/NoteSkins
%{_datadir}/stepmania-5.0/Themes
%{_datadir}/stepmania-5.0/Scripts
%{_datadir}/stepmania-5.0/Songs
%{_datadir}/icons/hicolor/*/apps/stepmania-ssc.png
%{_datadir}/icons/hicolor/scalable/apps/stepmania-ssc.svg
%{_datadir}/applications/stepmania.desktop

%changelog
* Wed Feb 10 2016 Naftuli Tzvi Kay <rfkrocktk@gmail.com> - 5.0.10-2
- Fix for required runtime libraries preventing installation.
* Wed Feb 10 2016 Naftuli Tzvi Kay <rfkrocktk@gmail.com> - 5.0.10-1
- Package StepMania. Everything looks great, but the GtkModule.so is missing a stack canary for some reason, and there's
  no binary in /usr/bin.
```

My patches:

FFMPEG add PIC compilation:
```
diff --git a/CMake/SetupFfmpeg.cmake b/CMake/SetupFfmpeg.cmake
index b700b67..910c9b2 100644
--- a/CMake/SetupFfmpeg.cmake
+++ b/CMake/SetupFfmpeg.cmake
@@ -20,6 +20,7 @@ list(APPEND FFMPEG_CONFIGURE
   "--disable-avfilter"
   "--disable-shared"
   "--enable-static"
+  "--enable-pic"
 )
 
 if(MACOSX)
```

Allow CMake building on i686:
```
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 61a1c95..7135e1a 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -295,7 +295,7 @@ else() # Linux
     if(CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64")
       message("Host processor is 64bit")
       sm_add_compile_definition("${SM_EXE_NAME}" CPU_X86_64)
-    elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "x86")
+    elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "x86" OR CMAKE_SYSTEM_PROCESSOR MATCHES "i686")
       message("Host processor is 32bit")
       sm_add_compile_definition("${SM_EXE_NAME}" CPU_X86)
     else()
```

Use system libraries for GLEW, jsoncpp, tomcrypt, tommath, libpng:
```
diff --git a/extern/CMakeLists.txt b/extern/CMakeLists.txt
index 46284d6..63eb255 100644
--- a/extern/CMakeLists.txt
+++ b/extern/CMakeLists.txt
@@ -1,6 +1,4 @@
 include(CMakeProject-lua.cmake)
-include(CMakeProject-glew.cmake)
-include(CMakeProject-json.cmake)
 if (APPLE OR MSVC)
   include(CMakeProject-mad.cmake)
   include(CMakeProject-zlib.cmake)
@@ -14,7 +12,4 @@ endif()
 if (NOT SYSTEM_PCRE_FOUND)
   include(CMakeProject-pcre.cmake)
 endif()
-include(CMakeProject-tomcrypt.cmake)
-include(CMakeProject-tommath.cmake)
-include(CMakeProject-png.cmake)
 
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 61a1c95..7a15d54 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -328,7 +328,7 @@ list(APPEND SMDATA_LINK_LIB
   "zlib"
   "jsoncpp"
   "png"
-  "glew"
+  "GLEW"
   "jpeg"
 )
```

The problem is that Lua should be statically compiled into the built binary and it isn't:

```
$ ldd /usr/share/stepmania-5.0/stepmania | grep lua
liblua-5.1.so => /lib64/liblua-5.1.so (0x00007f4c2f167000)
```

This causes StepMania to crash on startup:

```
$ /usr/share/stepmania-5.0/stepmania
/usr/share/stepmania-5.0/stepmania: symbol lookup error: /usr/share/stepmania-5.0/stepmania: undefined symbol: luaL_pushtype
```

For some reason, even though Fedora 23 provides Lua 5.1, it doesn't have that symbol in its defined types:

```
$ nm -D --defined-only /usr/lib64/liblua-5.1.so | grep -i luaL_push
00000000000189d0 T luaL_pushresult
```

How can I patch the CMake build for Lua to statically link the symbols into the StepMania binary?

**Comment by kyzentun** [comment]
System lua won't work because cmd is a special thing that was added to stepmania's bundled lua.  Without cmd, no themes will work.

**Comment by rfkrocktk** [comment]
@kyzentun That's fine, I don't want or need a system Lua. However, the CMake for 5.0.10 is broken and doesn't statically compile/link Lua into the StepMania binary, as seen in my `ldd` output above.

**Comment by rfkrocktk** [comment]
Interestingly enough, FFMPEG _is_ statically compiled into the build, but not Lua. I'm trying to figure out what's different about the CMake configuration for FFMPEG and apply that to Lua.

**Comment by rfkrocktk** [comment]
I'm pretty sure I've solved it with this:

```
diff --git a/extern/CMakeProject-lua.cmake b/extern/CMakeProject-lua.cmake
index 9da0c10..4fe3984 100644
--- a/extern/CMakeProject-lua.cmake
+++ b/extern/CMakeProject-lua.cmake
@@ -58,7 +58,7 @@ set(LUA_HPP
 source_group("" FILES ${LUA_SRC})
 source_group("" FILES ${LUA_HPP})
 
-add_library("lua-5.1" ${LUA_SRC} ${LUA_HPP})
+add_library("lua-5.1" STATIC ${LUA_SRC} ${LUA_HPP})
 
 set_property(TARGET "lua-5.1" PROPERTY FOLDER "External Libraries")
```

**Comment by rfkrocktk** [comment]
This is fixed by #1008.

**Comment by shakesoda** [comment]
@wolfman2000 naming library outputs the same thing as system libs is bad juju.
